File: flutter-voice-call-feature/lib/features/voice_call/services/voice_call_signaling_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart' show debugPrint;

import 'package:socket_io_client/socket_io_client.dart' as socket_io;

// Thin wrapper around the Socket.IO connection to the NestJS CallGateway
// (the `/calls` namespace). Only ever exchanges call setup/state and WebRTC
// signaling (offer/answer/ICE candidates) -- never audio.
//
// INTEGRATION POINT: `authToken` should be whatever token your app's
// existing authentication already produces (the same one your other API
// calls use). The server resolves it to a user id itself and never trusts a
// client-supplied id, so this class never needs to know your app's user id
// ahead of time.
class VoiceCallSignalingService {
  VoiceCallSignalingService({required this.baseUrl, required this.authToken});

  final String baseUrl;
  final String authToken;

  socket_io.Socket? _socket;

  final _onInvite = StreamController<Map<String, dynamic>>.broadcast();
  final _onAccept = StreamController<Map<String, dynamic>>.broadcast();
  final _onReject = StreamController<Map<String, dynamic>>.broadcast();
  final _onCancel = StreamController<Map<String, dynamic>>.broadcast();
  final _onHangup = StreamController<Map<String, dynamic>>.broadcast();
  final _onOffer = StreamController<Map<String, dynamic>>.broadcast();
  final _onAnswer = StreamController<Map<String, dynamic>>.broadcast();
  final _onIceCandidate = StreamController<Map<String, dynamic>>.broadcast();
  final _onFailed = StreamController<Map<String, dynamic>>.broadcast();
  final _onError = StreamController<Map<String, dynamic>>.broadcast();
  final _onConnectionChange = StreamController<bool>.broadcast();

  Stream<Map<String, dynamic>> get onInvite => _onInvite.stream;
  Stream<Map<String, dynamic>> get onAccept => _onAccept.stream;
  Stream<Map<String, dynamic>> get onReject => _onReject.stream;
  Stream<Map<String, dynamic>> get onCancel => _onCancel.stream;
  Stream<Map<String, dynamic>> get onHangup => _onHangup.stream;
  Stream<Map<String, dynamic>> get onOffer => _onOffer.stream;
  Stream<Map<String, dynamic>> get onAnswer => _onAnswer.stream;
  Stream<Map<String, dynamic>> get onIceCandidate => _onIceCandidate.stream;
  Stream<Map<String, dynamic>> get onFailed => _onFailed.stream;
  Stream<Map<String, dynamic>> get onError => _onError.stream;

  /// Emits `true`/`false` as the WebSocket connects/disconnects. A `false`
  /// while a call is active means the call can no longer be signaled --
  /// treat it as `VoiceCallError.webSocketDisconnected`.
  Stream<bool> get onConnectionChange => _onConnectionChange.stream;

  bool get isConnected => _socket?.connected ?? false;

  void connect() {
    final socket = socket_io.io(
      '$baseUrl/calls',
      socket_io.OptionBuilder()
          .setTransports(['websocket'])
          .setAuth({'token': authToken})
          .disableAutoConnect()
          .build(),
    );
    _socket = socket;

    socket.onConnect((_) {
      _log('connected');
      _onConnectionChange.add(true);
    });
    socket.onDisconnect((_) {
      _log('disconnected');
      _onConnectionChange.add(false);
    });
    socket.on('call:invite', (data) => _onInvite.add(_asMap(data)));
    socket.on('call:accept', (data) => _onAccept.add(_asMap(data)));
    socket.on('call:reject', (data) => _onReject.add(_asMap(data)));
    socket.on('call:cancel', (data) => _onCancel.add(_asMap(data)));
    socket.on('call:hangup', (data) => _onHangup.add(_asMap(data)));
    socket.on('webrtc:offer', (data) => _onOffer.add(_asMap(data)));
    socket.on('webrtc:answer', (data) => _onAnswer.add(_asMap(data)));
    socket.on('webrtc:ice-candidate', (data) => _onIceCandidate.add(_asMap(data)));
    socket.on('call:failed', (data) => _onFailed.add(_asMap(data)));
    socket.on('call:error', (data) => _onError.add(_asMap(data)));

    socket.connect();
  }

  Map<String, dynamic> _asMap(dynamic data) => Map<String, dynamic>.from(data as Map);

  void invite(String calleeId, void Function(Map<String, dynamic> ack) onAck) {
    _log('Calling user $calleeId');
    _socket?.emitWithAck('call:invite', {'calleeId': calleeId}, ack: (dynamic data) {
      onAck(_asMap(data));
    });
  }

  void accept(String callId) => _socket?.emit('call:accept', {'callId': callId});
  void reject(String callId) => _socket?.emit('call:reject', {'callId': callId});
  void cancel(String callId) => _socket?.emit('call:cancel', {'callId': callId});
  void hangup(String callId) => _socket?.emit('call:hangup', {'callId': callId});
  void notifyConnected(String callId) => _socket?.emit('call:connected', {'callId': callId});
  void notifyFailed(String callId) => _socket?.emit('call:failed', {'callId': callId});

  void sendOffer(String callId, Map<String, dynamic> sdp) {
    _log('Creating offer');
    _socket?.emit('webrtc:offer', {'callId': callId, 'sdp': sdp});
  }

  void sendAnswer(String callId, Map<String, dynamic> sdp) {
    _socket?.emit('webrtc:answer', {'callId': callId, 'sdp': sdp});
  }

  void sendIceCandidate(String callId, Map<String, dynamic> candidate) {
    _socket?.emit('webrtc:ice-candidate', {'callId': callId, 'candidate': candidate});
  }

  void _log(String message) => debugPrint('[CALL] $message');

  void dispose() {
    _socket?.dispose();
    _socket = null;
    _onInvite.close();
    _onAccept.close();
    _onReject.close();
    _onCancel.close();
    _onHangup.close();
    _onOffer.close();
    _onAnswer.close();
    _onIceCandidate.close();
    _onFailed.close();
    _onError.close();
    _onConnectionChange.close();
  }
}

File: web-voice-call-client/src/features/voiceCall/voiceCallSignalingService.ts
import { io, type Socket } from 'socket.io-client'
import type { IceCandidatePayload, SdpPayload } from './types'

interface InviteAck {
  callId?: string
  state?: string
}

interface CallEventPayload {
  callId: string
  from: string
}

interface ErrorPayload {
  reason?: string
  callId?: string
  calleeId?: string
}

// Thin wrapper around the Socket.IO connection to the NestJS CallGateway
// (the `/calls` namespace). Only ever exchanges call setup/state and WebRTC
// signaling (offer/answer/ICE candidates) -- never audio. Mirrors
// VoiceCallSignalingService from the Flutter feature 1:1, event for event.
//
// INTEGRATION POINT: `authToken` should be whatever token your app's real
// authentication produces. The server resolves it to a user id itself and
// never trusts a client-supplied id.
export class VoiceCallSignalingService {
  private socket: Socket | null = null

  onInvite?: (data: CallEventPayload) => void
  onAccept?: (data: CallEventPayload) => void
  onReject?: (data: CallEventPayload) => void
  onCancel?: (data: CallEventPayload) => void
  onHangup?: (data: CallEventPayload) => void
  onOffer?: (data: { callId: string; sdp: SdpPayload }) => void
  onAnswer?: (data: { callId: string; sdp: SdpPayload }) => void
  onIceCandidate?: (data: { callId: string; candidate: IceCandidatePayload }) => void
  onFailed?: (data: { callId: string }) => void
  onError?: (data: ErrorPayload) => void
  onConnectionChange?: (connected: boolean) => void

  constructor(
    private readonly baseUrl: string,
    private readonly authToken: string,
  ) {}

  get isConnected(): boolean {
    return this.socket?.connected ?? false
  }

  connect(): void {
    const socket = io(`${this.baseUrl}/calls`, {
      transports: ['websocket'],
      auth: { token: this.authToken },
      autoConnect: false,
    })
    this.socket = socket

    socket.on('connect', () => {
      this.log('connected')
      this.onConnectionChange?.(true)
    })
    socket.on('disconnect', () => {
      this.log('disconnected')
      this.onConnectionChange?.(false)
    })
    socket.on('call:invite', (data: CallEventPayload) => this.onInvite?.(data))
    socket.on('call:accept', (data: CallEventPayload) => this.onAccept?.(data))
    socket.on('call:reject', (data: CallEventPayload) => this.onReject?.(data))
    socket.on('call:cancel', (data: CallEventPayload) => this.onCancel?.(data))
    socket.on('call:hangup', (data: CallEventPayload) => this.onHangup?.(data))
    socket.on('webrtc:offer', (data: { callId: string; sdp: SdpPayload }) => this.onOffer?.(data))
    socket.on('webrtc:answer', (data: { callId: string; sdp: SdpPayload }) => this.onAnswer?.(data))
    socket.on('webrtc:ice-candidate', (data: { callId: string; candidate: IceCandidatePayload }) =>
      this.onIceCandidate?.(data),
    )
    socket.on('call:failed', (data: { callId: string }) => this.onFailed?.(data))
    socket.on('call:error', (data: ErrorPayload) => this.onError?.(data))

    socket.connect()
  }

  invite(calleeId: string, onAck: (ack: InviteAck) => void): void {
    this.log(`Calling user ${calleeId}`)
    this.socket?.emit('call:invite', { calleeId }, (ack: InviteAck) => onAck(ack))
  }

  accept(callId: string): void {
    this.socket?.emit('call:accept', { callId })
  }

  reject(callId: string): void {
    this.socket?.emit('call:reject', { callId })
  }

  cancel(callId: string): void {
    this.socket?.emit('call:cancel', { callId })
  }

  hangup(callId: string): void {
    this.socket?.emit('call:hangup', { callId })
  }

  notifyConnected(callId: string): void {
    this.socket?.emit('call:connected', { callId })
  }

  notifyFailed(callId: string): void {
    this.socket?.emit('call:failed', { callId })
  }

  sendOffer(callId: string, sdp: SdpPayload): void {
    this.log('Creating offer')
    this.socket?.emit('webrtc:offer', { callId, sdp })
  }

  sendAnswer(callId: string, sdp: SdpPayload): void {
    this.socket?.emit('webrtc:answer', { callId, sdp })
  }

  sendIceCandidate(callId: string, candidate: IceCandidatePayload): void {
    this.socket?.emit('webrtc:ice-candidate', { callId, candidate })
  }

  private log(message: string): void {
    console.log(`[CALL] ${message}`)
  }

  dispose(): void {
    this.socket?.disconnect()
    this.socket = null
  }
}
